ntoskrnl/io/iomgr/deviface.c

```diff
@@ -30,185 +31,421 @@ static PWCHAR BaseKeyString = L"\\Registry\\Machine\\System\\CurrentControlSet\\
 
 static
 NTSTATUS
-OpenRegistryHandlesFromSymbolicLink(IN PUNICODE_STRING SymbolicLinkName,
-                                    IN ACCESS_MASK DesiredAccess,
-                                    IN OPTIONAL PHANDLE GuidKey,
-                                    IN OPTIONAL PHANDLE DeviceKey,
-                                    IN OPTIONAL PHANDLE InstanceKey)
+IopBuildSymbolicLink(_In_ PUNICODE_STRING PrefixString,
+                     _In_ PUNICODE_STRING DeviceString,
+                     _In_ PUNICODE_STRING GuidString,
+                     _In_opt_ PUNICODE_STRING ReferenceString,
+                     _Out_ PUNICODE_STRING SymbolicLinkName)
 {
-    OBJECT_ATTRIBUTES ObjectAttributes;
-    UNICODE_STRING BaseKeyU;
-    UNICODE_STRING GuidString, SubKeyName, ReferenceString;
-    PWCHAR StartPosition, EndPosition;
-    HANDLE ClassesKey;
-    PHANDLE GuidKeyRealP, DeviceKeyRealP, InstanceKeyRealP;
-    HANDLE GuidKeyReal, DeviceKeyReal, InstanceKeyReal;
+    UNICODE_STRING SymbolicLink;
     NTSTATUS Status;
 
-    SubKeyName.Buffer = NULL;
+    /* Build up new symolic link */
+    SymbolicLink.Length = 0;
+    SymbolicLink.MaximumLength = PrefixString->Length + DeviceString->Length + sizeof(L"#") + GuidString->Length + ReferenceString->Length + sizeof(UNICODE_NULL);
```

Comment:
You unconditionally dereference `ReferenceString` here, so you'll crash if it's null.

Reply:
Indeed. But how is better to specify the condition that will allow to choose the proper value depending on reference string? I already tried `ReferenceString != NULL ? ReferenceString->Length : sizeof(L"\\")`, but it makes output symbolic link allocated incorrectly even if refstring isn't `NULL`. So, as result, it becomes empty. What's wrong with my condition? :thinking:

Reply:
What form should the symbolic link have? (btw. the comment has a typo: it's "sym*b*olic")

Reply:
Your condition looks fine (other than `sizeof(L"\\")` is 4, when what you want is probably 2). Did you have it in parentheses? `?` has lower precedence than `+`.

Reply:
> What form should the symbolic link have? (btw. the comment has a typo: it's "sym_b_olic")

Something like this: `\??\ACPI#PNP0501#1#{GUID}\ReferenceString`.
It starts from a prefix (kernel-mode `\??\` or user-mode `\\?\`), then goes a device string (like that `ACPI#PNP0501#1#`), GUID (in `{}` brackets) and ends by a reference string (`\SomeCondition` or just `\` if it's empty). :slightly_smiling_face:

Reply:
OK, thanks. Added parentheses around that condition. It should fix the problem properly. :slightly_smiling_face: